wgpu/backends/rs.py

```diff
@@ -149,8 +149,47 @@ def get_surface_id_from_canvas(canvas):
         #         [ns_window.contentView setLayer:metal_layer];
         #         surface = wgpu_create_surface_from_metal_layer(metal_layer);
         #     }
-        layer = ffi.cast("void *", win_id)
-        return _lib.wgpu_create_surface_from_metal_layer(layer)
+        window = ctypes.c_void_p(win_id)
+
+        objc = ctypes.cdll.LoadLibrary(ctypes.util.find_library("objc"))
+        objc.objc_getClass.restype = ctypes.c_void_p
+        objc.sel_registerName.restype = ctypes.c_void_p
+        objc.objc_msgSend.restype = ctypes.c_void_p
+        objc.objc_msgSend.argtypes = [ctypes.c_void_p, ctypes.c_void_p]
+
+        content_view_sel = objc.sel_registerName(b"contentView")
+        set_wants_layer_sel = objc.sel_registerName(b"setWantsLayer:")
+        responds_to_sel_sel = objc.sel_registerName(b"respondsToSelector:")
+        layer_sel = objc.sel_registerName(b"layer")
+        set_layer_sel = objc.sel_registerName(b"setLayer:")
+
+        # Try some duck typing to see what kind of object the window pointer points to
+        # Qt doesn't return a NSWindow, but a QNSView instead, which is subclass of NSView.
+        if objc.objc_msgSend(
+            window, responds_to_sel_sel, ctypes.c_void_p(content_view_sel)
+        ):
+            # NSWindow instances respond to contentView selector
+            content_view = objc.objc_msgSend(window, content_view_sel)
+        elif objc.objc_msgSend(window, responds_to_sel_sel, ctypes.c_void_p(layer_sel)):
+            # NSView instances respond to layer selector
+            # Let's assume that the given window pointer is actually the content view
+            content_view = window
+        else:
+            # Can't tell what kind of object window is
+            raise TypeError()
```

Comment:
Let's do something like `raise TypeError("bla bla fail")`

Reply:
Actually make it a `RuntimeError`; I think that's more appropriate from Python's point of view.

Reply:
Changed the TypeError to RuntimeError, and I've added a message. Do you think this is enough info?

Reply:
Perfect!